execution/nodeInterface/NodeInterface.go

```diff
@@ -53,90 +51,126 @@ var merkleTopic common.Hash
-			batch = 0
-		} else if errors.Is(err, blockAfterLatestBatch) {
-			return 0, nil
-		} else {
-			return 0, err
-		}
+		return 0, err
+	}
+	// batches not yet posted have 0 confirmations but no error
+	batchNum, found, err := n.msgNumToInboxBatch(msgNum)
+	if err != nil || !found {
+		return 0, err
 	}
-	meta, err := node.InboxTracker.GetBatchMetadata(batch)
+	parentChainBlockNum, err := node.ExecEngine.GetBatchFetcher().GetBatchParentChainBlock(batchNum)
 	if err != nil {
 		return 0, err
 	}
-	if node.L1Reader.IsParentChainArbitrum() {
-		parentChainClient := node.L1Reader.Client()
+
+	if node.ParentChainReader.IsParentChainArbitrum() {
+		parentChainClient := node.ParentChainReader.Client()
 		parentNodeInterface, err := node_interfacegen.NewNodeInterface(types.NodeInterfaceAddress, parentChainClient)
 		if err != nil {
 			return 0, err
 		}
-		parentChainBlock, err := parentChainClient.BlockByNumber(n.context, new(big.Int).SetUint64(meta.ParentChainBlock))
+		parentChainBlock, err := parentChainClient.BlockByNumber(n.context, new(big.Int).SetUint64(parentChainBlockNum))
 		if err != nil {
 			// Hide the parent chain RPC error from the client in case it contains sensitive information.
 			// Likely though, this error is just "not found" because the block got reorg'd.
-			return 0, fmt.Errorf("failed to get parent chain block %v containing batch", meta.ParentChainBlock)
+			return 0, fmt.Errorf("failed to get parent chain block %v containing batch", parentChainBlockNum)
 		}
 		confs, err := parentNodeInterface.GetL1Confirmations(&bind.CallOpts{Context: n.context}, parentChainBlock.Hash())
 		if err != nil {
 			log.Warn(
 				"Failed to get L1 confirmations from parent chain",
-				"blockNumber", meta.ParentChainBlock,
+				"blockNumber", parentChainBlockNum,
 				"blockHash", parentChainBlock.Hash(), "err", err,
 			)
 			return 0, fmt.Errorf("failed to get L1 confirmations from parent chain for block %v", parentChainBlock.Hash())
 		}
 		return confs, nil
 	}
-	latestL1Block, latestBatchCount := node.InboxReader.GetLastReadBlockAndBatchCount()
-	if latestBatchCount <= batch {
-		return 0, nil // batch was reorg'd out?
-	}
-	if latestL1Block < meta.ParentChainBlock || arbutil.BlockNumberToMessageCount(blockNum, genesis) > meta.MessageCount {
+	if node.ParentChainReader == nil {
 		return 0, nil
 	}
-	canonicalHash := bc.GetCanonicalHash(header.Number.Uint64())
-	if canonicalHash != header.Hash() {
-		return 0, errors.New("block hash is non-canonical")
+	latestHeader, err := node.ParentChainReader.LastHeaderWithError()
+	if err != nil {
+		return 0, err
+	}
+	if latestHeader == nil {
+		return 0, errors.New("no headers read from l1")
+	}
+	latestBlockNum := latestHeader.Number.Uint64()
+	if latestBlockNum < parentChainBlockNum {
+		return 0, nil
 	}
-	confs := (latestL1Block - meta.ParentChainBlock) + 1 + node.InboxReader.GetDelayBlocks()
-	return confs, nil
+	return (latestBlockNum - parentChainBlockNum), nil
```

Comment:
The goal of `GetLastReadBlockAndBatchCount` was to be safe against reorgs. I think if a reorg appeared that added a bunch of blocks at once (this seems possible in eth PoS), this method might say that a batch has a bunch of confs if the inbox reader hasn't picked up the reorg yet.

That said, people shouldn't be looking at # of blocks as a confirmation metric post-PoS anyways, so I'm fine deprecating this if it isn't possible to get both atomically.